programs/metadata/src/lib.rs

```diff
@@ -0,0 +1,195 @@
+                ErrorCode::InvalidOperationInCurrentSlot
+            );
+            item.value = new_value;
+            item.last_updated_slot = current_slot;
+            metadata.last_updated_slot = current_slot;
+        } else {
+            msg!("Key: {}", key);
+            return Err(error!(ErrorCode::KeyNotFound));
+        }
+        Ok(())
+    }
+
+    // Instruction to append data to an existing MetadataItem
+    pub fn append_metadata_item(
+        ctx: Context<ModifyMetadata>,
+        key: String,
+        additional_value: Vec<u8>,
+    ) -> Result<()> {
+        let metadata = &mut ctx.accounts.metadata;
+        let current_slot = Clock::get()?.slot;
+        if let Some(item) = metadata.items.iter_mut().find(|item| item.key == key) {
+            require_gt!(
+                current_slot,
+                item.last_updated_slot,
+                ErrorCode::InvalidOperationInCurrentSlot
+            );
+            item.value.extend(additional_value);
+            item.last_updated_slot = current_slot;
+            metadata.last_updated_slot = current_slot;
+        } else {
+            msg!("Key: {}", key);
+            return Err(error!(ErrorCode::KeyNotFound));
+        }
+        Ok(())
+    }
+}
+
+#[derive(Accounts)]
+pub struct CreateMetadata<'info> {
+    #[account(init, payer = payer, seeds = [dao_treasury.key().as_ref()], bump, space = 8 + DEFAULT_SPACE)]
+    pub metadata: Account<'info, Metadata>,
+    pub dao_treasury: Signer<'info>,
+    #[account(mut)]
+    pub payer: Signer<'info>,
+    /// CHECK: This is the metadata delegate account, it only ever signs
+    pub delegate: UncheckedAccount<'info>,
+    pub system_program: Program<'info, System>,
+}
+
+#[derive(Accounts)]
+pub struct IncreaseMetadataAccountSize<'info> {
+    #[account(
+        mut,
+        has_one = delegate,
+        realloc = metadata.to_account_info().data_len() + INCREASE_IN_SPACE, // TODO: Set the max size, after which heap memory breaks
+        realloc::payer = payer,
+        realloc::zero = false,
+    )]
+    pub metadata: Account<'info, Metadata>,
+    pub delegate: Signer<'info>,
+    #[account(mut)]
+    pub payer: Signer<'info>,
+    pub system_program: Program<'info, System>,
+}
+
+#[derive(Accounts)]
+pub struct SetDelegate<'info> {
+    #[account(mut, has_one = dao_treasury)]
+    pub metadata: Account<'info, Metadata>,
+    pub dao_treasury: Signer<'info>,
+    /// CHECK: This is the metadata delegate account, it only ever signs
+    pub new_delegate: UncheckedAccount<'info>,
+}
+
+#[derive(Accounts)]
+pub struct ModifyMetadata<'info> {
+    #[account(mut, has_one = delegate)]
+    pub metadata: Account<'info, Metadata>,
+    pub delegate: Signer<'info>,
+}
+
+#[error_code]
+pub enum ErrorCode {
+    #[msg("The provided metadata key already exists.")]
+    DuplicateKey,
+    #[msg("Operation cannot be performed on an item updated in the current slot.")]
+    InvalidOperationInCurrentSlot,
```

Comment:
This is something nallok and realms dao people mentioned. You don't want there to be multiple updates of a metadata object in the same slot because of security or something about spoofing changes? Plus if there's only one change per item per slot it makes going back and getting the history much cleaner.